### PR TITLE
[IPSCCP] Variable not visible at Og.

### PR DESCRIPTION
**Pull request by CarlosAlbertoEnciso**
https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not create a constant expression for the initializer value.

### CODE REVIEW

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 1]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
       StoreInst *SI = cast<StoreInst>(GV->user_back());
       SI->eraseFromParent();
     }
+
+    // Try to create a debug constant expression for the glbal variable
```
Comment (jryans) [inline llvm/lib/Transforms/IPO/SCCP.cpp:376]:
```suggestion
    // Try to create a debug constant expression for the global variable
```

**Review by jryans — APPROVED** [review]
This looks good to me overall, thanks for working on this improvement! 😄 

I noticed the PR title ends with a ":", perhaps some further words were lost or the ":" should be removed...?

I recommend waiting a few days for others to take a look as well.

**Comment by nikic** [comment]
I don't think this chunk of code for converting a Constant into a DIExpression belongs into SCCP.cpp.

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 2]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
...
+
+      // Create integer constant expression.
+      auto createIntExpression = [&DIB](const Constant *CV) -> DIExpression * {
+        const APInt &API = dyn_cast<ConstantInt>(CV)->getValue();
+        std::optional<uint64_t> InitIntOpt;
+        if (API.isNonNegative())
+          InitIntOpt = API.tryZExtValue();
+        else if (auto Temp = API.trySExtValue(); Temp.has_value())
+          // Transform a signed optional to unsigned optional.
+          InitIntOpt = (uint64_t)Temp.value();
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/IPO/SCCP.cpp:391]: let's avoid C-style casts!

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 3]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
...
+        return DIB.createConstantValueExpression(InitIntOpt.value());
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/IPO/SCCP.cpp:392]: This is unconditionally calling `std::optional::value`, but the control flow above makes it look like we may have a `nullopt` here. This would be even more evident if we had not used an uninitialized variable here (it would be evident that we are missing an initialization case).

I think we need to account for the possibility of failure here:

```
DIExpression* createIntExpression(DIB, CV) {
        const APInt &API = dyn_cast<ConstantInt>(CV)->getValue();
        if (API.isNonNegative()) {
           if(optional<uint64_t> MaybeZExt = API.tryZExtValue())
                   return DIB.createConstantValueExpression(*MaybeZExt);
        }
        else if (optional<uint64_t> MaybeSExt = API.tryZExtValue())
                          return DIB.createConstantValueExpression(* MaybeSExt);
       return nullptr;
};

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 4]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
...
+      };
+
+      const Constant *CV = GV->getInitializer();
+      Type *Ty = GV->getValueType();
+      if (Ty->isIntegerTy()) {
+        GVEs[0]->replaceOperandWith(1, createIntExpression(CV));
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/IPO/SCCP.cpp:398]: Are we allowed to mutate a metadata like this? IIUC we can only do this if it is uniqued? Or not uniqued? I can never remember exactly

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 5]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
...
+      } else if (Ty->isFloatTy() || Ty->isDoubleTy()) {
+        const APFloat &APF = dyn_cast<ConstantFP>(CV)->getValueAPF();
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/IPO/SCCP.cpp:400]: a `dyn_cast` that is never checked should be a `cast`

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 6]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
...
+              const Value *V = CE->getOperand(0);
+              const Constant *CV = dyn_cast<Constant>(V);
+              if (CV && !isa<GlobalValue>(CV))
+                if (const ConstantInt *CI = dyn_cast<ConstantInt>(CV))
+                  GVEs[0]->replaceOperandWith(1, createIntExpression(CI));
+            }
+          }
+        }
+      }
+    }
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/IPO/SCCP.cpp:419]: The control flow here is a bit difficult to follow, but I think we will improve this a lot once we make the helper function and replace some of those nested `if`s with early returns.

**Review by felipepiovezan — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 3 above)
Comment (felipepiovezan) [reply llvm/lib/Transforms/IPO/SCCP.cpp:392]: Oh actually here we have another case where we should use `cast<ConstantInt>(CV)` instead of `dyn_cast`

**Comment by CarlosAlbertoEnciso** [comment]
@jryans, @felipepiovezan, @nikic: Thanks for your comments.
I will address them and upload a new patch.

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 2 above)
Comment (dwblaikie) [inline llvm/lib/Transforms/IPO/SCCP.cpp:391]: Probably also favor `*` rather than `.value()` - not sure if we've clarified that in the style guide or not...

Oh, and the if condition could be simpler - no need to use `has_value` at all:
```
else if (auto Temp = API.trySExtValue())
```
Should suffice?

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 6 above)
Comment (dwblaikie) [reply llvm/lib/Transforms/IPO/SCCP.cpp:419]: This could also be collapsed:
```
        } else {
          if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(CV)) {
            if (CE->getNumOperands() == 1) {
```
into this:
```
        } else if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(CV); CE->getNumOperands() == 1) {
```

But other refactoring might render that moot/be an alternative.

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 7]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
       StoreInst *SI = cast<StoreInst>(GV->user_back());
       SI->eraseFromParent();
     }
+
+    // Try to create a debug constant expression for the glbal variable
+    // initializer value.
+    SmallVector<DIGlobalVariableExpression *, 1> GVEs;
+    GV->getDebugInfo(GVEs);
+    if (GVEs.size() == 1) {
```
[the remark this answers is not in the document]
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:380]: Good point.

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 1 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:376]: Fixed the typo.

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 8]
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
       StoreInst *SI = cast<StoreInst>(GV->user_back());
       SI->eraseFromParent();
     }
+
+    // Try to create a debug constant expression for the glbal variable
+    // initializer value.
```
[the remark this answers is not in the document]
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:377]: Moving the code to a helper function.

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 2 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:391]: Using instead `static_cast<uint64_t>`

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 3 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:392]: Changed to use `cast`.

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 4 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:398]: The function `replaceOperandWith` takes that into consideration.

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 5 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:400]: Changed to `cast`

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 6 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:419]: Replaced with your suggested change.

**Comment by CarlosAlbertoEnciso** [comment]
@jryans, @felipepiovezan, @nikic, @dwblaikie:
Thanks for all your feedback. I have uploaded a new patch that addresses all your points.
I am sorry if there are extra commits on this branch. This is the first update since moving to GitHub.

**Review by jryans — APPROVED** [review]
Looks good to me, thanks! 😄 

Other reviewers had more comments last time around, so probably best to wait for at least one more review before landing.

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 9]
```diff
@@ -106,14 +107,71 @@ static void findReturnsToZap(Function &F,
...
+
+  if (isa<ConstantPointerNull>(CV)) {
+    GVEs[0]->replaceOperandWith(1, DIB.createConstantValueExpression(0));
+    return;
+  }
+  if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(CV);
+      CE->getNumOperands() == 1) {
+    const Value *V = CE->getOperand(0);
+    const Constant *CV = dyn_cast<Constant>(V);
+    if (CV && !isa<GlobalValue>(CV);
```
Comment (dwblaikie) [inline llvm/lib/Transforms/IPO/SCCP.cpp:158]: I'm not sure this does what's intended (assuming the intent is that the `CV && !isa` part is meant to be part of the boolean condition of the `if` as well as the `CI` init) - these boolean expressions would have no effect, right? (the second part of the if, after the semicolon, would be the only condition?

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 10]
```diff
@@ -106,14 +107,71 @@ static void findReturnsToZap(Function &F,
...
+  }
+
+  if (!Ty->isPointerTy())
+    return;
+
+  if (isa<ConstantPointerNull>(CV)) {
+    GVEs[0]->replaceOperandWith(1, DIB.createConstantValueExpression(0));
+    return;
+  }
+  if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(CV);
```
Comment (dwblaikie) [inline llvm/lib/Transforms/IPO/SCCP.cpp:154]: This is probably problematic too - the initializer part of the `if` isn't part of the boolean condition - so the `CE` would be unconditionally dereferenced in the `CE->getNumOperands()` call - so either the `CE` init should use `cast`, not `dyn_cast` - or the condition needs to be `CE && CE->getNumOperands() ...` (or nested `if`s)

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 9 above)
Comment (felipepiovezan) [reply llvm/lib/Transforms/IPO/SCCP.cpp:158]: Yeah, I believe the intent here would be:

```
if (auto CI = dyn_cast_or_null<ConstantInt>(V)) {
  ...
}
```

Note that `!isa<GlobalValue>(CV)` is not necessary: a global value is always a pointer, therefore it cannot be a `ConstantInt`

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 10 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:154]: Good catch. Replaced with nested `if`s.

**Comment by nikic** [comment]
I still think that the logic for "convert Constant into DIExpression" should not be part of IPSCCP, but be a helper in some generic location (not sure where, debuginfo folks would know). This seems generically useful (to the point that I'm surprised it does not exist yet).

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 11]
```diff
@@ -106,14 +107,71 @@ static void findReturnsToZap(Function &F,
...
+
+  // Create integer constant expression.
+  auto createIntegerExpression = [&DIB](const Constant *CV) -> DIExpression * {
+    const APInt &API = cast<ConstantInt>(CV)->getValue();
+    std::optional<uint64_t> InitIntOpt;
+    if (API.isNonNegative())
+      InitIntOpt = API.tryZExtValue();
+    else if (auto Temp = API.trySExtValue())
+      // Transform a signed optional to unsigned optional.
+      InitIntOpt = static_cast<uint64_t>(*Temp);
```
[the remark this answers is not in the document]
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:126]: Good point. I think the original code would be better if we are dealing with `APSInt`. But for this case, we have `APInt`.
Changed to 
```
    std::optional<int64_t> InitIntOpt = API.trySExtValue();
    return InitIntOpt ? DIB.createConstantValueExpression(
                            static_cast<uint64_t>(*InitIntOpt))
                      : nullptr;
```

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 9 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:158]: Good catch and nice simplification.

**Comment by jryans** [comment]
> I still think that the logic for "convert Constant into DIExpression" should not be part of IPSCCP, but be a helper in some generic location (not sure where, debuginfo folks would know). This seems generically useful (to the point that I'm surprised it does not exist yet).

I agree, it does seem good to make this available as a shared helper. Perhaps it could go to https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/Local.cpp? There are various shared debug info helpers stashed there already. (At some stage, a separate file just for debug info utils might be a good idea, but that can be separate task.)

**Comment by CarlosAlbertoEnciso** [comment]
Uploaded new patch with the new logic moved to https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/Local.cpp

File: llvm/lib/Transforms/IPO/SCCP.cpp [hunk 12]
```diff
@@ -106,14 +107,14 @@ static void findReturnsToZap(Function &F,
...
 
-static bool runIPSCCP(
-    Module &M, const DataLayout &DL, FunctionAnalysisManager *FAM,
-    std::function<const TargetLibraryInfo &(Function &)> GetTLI,
-    std::function<TargetTransformInfo &(Function &)> GetTTI,
-    std::function<AssumptionCache &(Function &)> GetAC,
-    std::function<DominatorTree &(Function &)> GetDT,
-    std::function<BlockFrequencyInfo &(Function &)> GetBFI,
-    bool IsFuncSpecEnabled) {
+static bool
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/IPO/SCCP.cpp:110]: this whitespace diff should not be part of the patch

File: llvm/include/llvm/Transforms/Utils/Local.h [hunk 13]
```diff
@@ -515,6 +515,14 @@ Value *invertCondition(Value *Condition);
 /// function, explicitly materialize the maximal set in the IR.
 bool inferAttributesFromOthers(Function &F);
 
+///===---------------------------------------------------------------------===//
+///  DebugInfo utilities
```
Comment (jryans) [inline llvm/include/llvm/Transforms/Utils/Local.h:519]: I would think it fits in with the existing "Dbg Intrinsic utilities" section:

https://github.com/llvm/llvm-project/blob/df8e0d057d67a8167f2bdc81d8b61033beb55da5/llvm/include/llvm/Transforms/Utils/Local.h#L258

Perhaps it's better to append it to the end of that section instead of creating a new, similar section?

File: llvm/lib/Transforms/Utils/Local.cpp [hunk 14]
```diff
@@ -3606,3 +3606,41 @@ bool llvm::inferAttributesFromOthers(Function &F) {
 
   return Changed;
 }
+
+DIExpression *llvm::getExpressionForConstant(DIBuilder &DIB, const Constant *C,
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/Utils/Local.cpp:3610]: We never check if `C` or `Ty` are null, we can make these references

**Review by felipepiovezan — APPROVED** [review]
I left two other minor comments, otherwise the patch LGTM!

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 12 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp:110]: Removed.

File: llvm/lib/Transforms/Utils/Local.cpp (hunk 14 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/Utils/Local.cpp:3610]: Passing them as references.

**Comment by CarlosAlbertoEnciso** [comment]
@jryans, @felipepiovezan: uploaded new patch to reflect your latest comments.
@nikic, @dwblaikie: do you have any additional comments?

File: llvm/lib/Transforms/Utils/Local.cpp [hunk 15]
```diff
@@ -3074,6 +3074,44 @@ void llvm::hoistAllInstructionsInto(BasicBlock *DomBlock, Instruction *InsertPt,
...
+  if (!Ty.isPointerTy())
+    return nullptr;
+
+  if (isa<ConstantPointerNull>(C))
+    return DIB.createConstantValueExpression(0);
+
+  if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(&C)) {
+    if (CE->getNumOperands() == 1) {
+      const Value *V = CE->getOperand(0);
+      if (auto CI = dyn_cast_or_null<ConstantInt>(V)) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/Local.cpp:3107]: Omit braces

File: llvm/lib/Transforms/Utils/Local.cpp [hunk 16]
```diff
@@ -3074,6 +3074,44 @@ void llvm::hoistAllInstructionsInto(BasicBlock *DomBlock, Instruction *InsertPt,
...
+  }
+
+  if (!Ty.isPointerTy())
+    return nullptr;
+
+  if (isa<ConstantPointerNull>(C))
+    return DIB.createConstantValueExpression(0);
+
+  if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(&C)) {
+    if (CE->getNumOperands() == 1) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/Local.cpp:3105]:
```suggestion
    if (CE->getOpcode() == Instruction::IntToPtr) {
```
I *think* this is what you're actually trying to handle here? I think this would be clearer than identifying inttoptr as "an expression with one operand that has pointer result and ConstantInt operand".

File: llvm/test/Transforms/SCCP/pr50901.ll [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+  %4 = load i8, ptr @_ZL4g_55, align 1, !dbg !64, !tbaa !65, !range !67, !noundef !68
+  %tobool = trunc i8 %4 to i1, !dbg !64
+  %frombool = zext i1 %tobool to i8, !dbg !69
+  store i8 %frombool, ptr @g_5, align 1, !dbg !69, !tbaa !65
+  %5 = load ptr, ptr @_ZL4g_66, align 8, !dbg !70, !tbaa !71
+  store ptr %5, ptr @g_6, align 8, !dbg !73, !tbaa !71
+  %6 = load ptr, ptr @_ZL4g_77, align 8, !dbg !74, !tbaa !71
+  store ptr %6, ptr @g_7, align 8, !dbg !75, !tbaa !71
+  ret void, !dbg !76
+}
```
Comment (nikic) [inline llvm/test/Transforms/SCCP/pr50901.ll:90]: You should be able to drop all the `!tbaa` metadata (really, any non-dbg metadata) from this test. It shouldn't be relevant for IPSCCP.

File: llvm/test/Transforms/SCCP/pr50901.ll [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+; CHECK-DAG: ![[DBG4]] = distinct !DIGlobalVariable(name: "g_44", {{.*}}
+; CHECK: ![[G5:[0-9]+]] = !DIGlobalVariableExpression(var: ![[DBG5:[0-9]+]], expr: !DIExpression(DW_OP_constu, 1, DW_OP_stack_value))
+; CHECK-DAG: ![[DBG5]] = distinct !DIGlobalVariable(name: "g_55", {{.*}}
+; CHECK: ![[G6:[0-9]+]] = !DIGlobalVariableExpression(var: ![[DBG6:[0-9]+]], expr: !DIExpression(DW_OP_constu, 0, DW_OP_stack_value))
+; CHECK-DAG: ![[DBG6]] = distinct !DIGlobalVariable(name: "g_66", {{.*}}
+; CHECK: ![[G7:[0-9]+]] = !DIGlobalVariableExpression(var: ![[DBG7:[0-9]+]], expr: !DIExpression(DW_OP_constu, 70, DW_OP_stack_value))
+; CHECK-DAG: ![[DBG7]] = distinct !DIGlobalVariable(name: "g_77", {{.*}}
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-linux-gnu"
```
Comment (nikic) [inline llvm/test/Transforms/SCCP/pr50901.ll:54]: Data layout and triple shouldn't be necessary. (If you specify a triple, it's either useless or you need to add REQUIRES, otherwise you will break the build.)

**Review by nikic — COMMENTED** [review]
Looks fine, just some nits.

File: llvm/test/Transforms/SCCP/pr50901.ll [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+@_ZL4g_22 = internal global float 0x4016333340000000, align 4, !dbg !27
+@_ZL4g_33 = internal global i8 98, align 1, !dbg !29
+@_ZL4g_44 = internal global i32 5, align 4, !dbg !31
+@_ZL4g_55 = internal global i8 1, align 1, !dbg !33
+@_ZL4g_66 = internal global ptr null, align 8, !dbg !35
+@_ZL4g_77 = internal global ptr inttoptr (i64 70 to ptr), align 8, !dbg !37
+
+define dso_local void @_Z3barv() !dbg !46 {
+entry:
+  %0 = load i32, ptr @_ZL4g_11, align 4, !dbg !49, !tbaa !50
```
Comment (djtodoro) [inline llvm/test/Transforms/SCCP/pr50901.ll:73]: I believe that you could use only one `!dbg` for all the instructions in the function and get rid of all the others.

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 10 above)
Comment (dwblaikie) [reply llvm/lib/Transforms/IPO/SCCP.cpp]: This and the other conditions seemed pretty broken, glad they're now fixed - but is there some test coverage that was missed and would've demonstrated these problems/should be added to ensure comprehensive test coverage?

File: llvm/test/Transforms/SCCP/pr50901.ll (hunk 18 above)
Comment (CarlosAlbertoEnciso) [reply llvm/test/Transforms/SCCP/pr50901.ll:54]: Removed `datalayout` and `triple`.

File: llvm/lib/Transforms/IPO/SCCP.cpp (hunk 10 above)
Comment (CarlosAlbertoEnciso) [reply llvm/lib/Transforms/IPO/SCCP.cpp]: I am happy to add some tests in [LocalTest.cpp](llvm/unittests/Transforms/Utils/LocalTest.cpp) to ensure we have test coverage for the new function.

File: llvm/test/Transforms/SCCP/pr50901.ll (hunk 17 above)
Comment (CarlosAlbertoEnciso) [reply llvm/test/Transforms/SCCP/pr50901.ll:90]: Removed all the `!tbaa` metadata.

**Comment by CarlosAlbertoEnciso** [comment]
Thanks for all your comments.
Updated to address:
@dwblaikie: include unittests (Covering the most common types for constants).
@djtodoro and @nikic: reduce the lit test (Removing redundant `!dbg`).
@nikic: extra braces and different `if` condition for pointer types.

**Comment by dwblaikie** [comment]
(nothing further from me, but I'll leave it to other folks more familiar with this to approve)

**Comment by CarlosAlbertoEnciso** [comment]
> (nothing further from me, but I'll leave it to other folks more familiar with this to approve)

Thanks

**Comment by CarlosAlbertoEnciso** [comment]
Thanks to all reviewers for your valuable comments.
Any additional points to address, before I merge the changes.

**Comment by nikic** [comment]
I think you can go ahead and merge this. My comments have been addressed, and other people already approved this previously.

**Comment by Endilll** [comment]
Looks like this breaks OpenCL: https://github.com/llvm/llvm-project/issues/50901#issuecomment-1870283804
